lib/wasi-types/src/wasi/wasix_manual.rs

```diff
@@ -0,0 +1,187 @@
+};
+
+pub type WasiHash = u128;
+pub type WasiSmallHash = u64;
+
+/// Thread local key
+pub type TlKey = u32;
+/// Thread local value
+pub type TlVal = u64;
+/// Thread local user data (associated with the value)
+pub type TlUser = u64;
+/// Long size used by checkpoints
+pub type Longsize = u64;
+
+/// The contents of a `subscription`, snapshot0 version.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union Snapshot0SubscriptionUnion {
+    pub clock: Snapshot0SubscriptionClock,
+    pub fd_readwrite: SubscriptionFsReadwrite,
+}
+/// The contents of a `subscription`.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union SubscriptionUnion {
+    pub clock: SubscriptionClock,
+    pub fd_readwrite: SubscriptionFsReadwrite,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Snapshot0Subscription {
+    pub userdata: Userdata,
+    pub type_: Eventtype,
+    pub u: Snapshot0SubscriptionUnion,
+}
+impl core::fmt::Debug for Snapshot0Subscription {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Snapshot0Subscription")
+            .field("userdata", &self.userdata)
+            .field("type", &self.type_)
+            .finish()
+    }
+}
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Subscription {
+    pub userdata: Userdata,
+    pub type_: Eventtype,
+    pub data: SubscriptionUnion,
+}
+impl core::fmt::Debug for Subscription {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Subscription")
+            .field("userdata", &self.userdata)
+            .field("type", &self.type_)
+            .finish()
+    }
+}
+
+impl From<Snapshot0Subscription> for Subscription {
+    fn from(other: Snapshot0Subscription) -> Self {
+        Self {
+            userdata: other.userdata,
+            type_: other.type_,
+            data: match other.type_ {
+                Eventtype::Clock => SubscriptionUnion {
+                    clock: unsafe {
+                        SubscriptionClock {
+                            clock_id: other.u.clock.id.into(),
+                            timeout: other.u.clock.timeout,
+                            precision: other.u.clock.precision,
+                            flags: other.u.clock.flags,
+                        }
+                    },
+                },
+                Eventtype::FdRead => SubscriptionUnion {
+                    fd_readwrite: unsafe { other.u.fd_readwrite },
+                },
+                Eventtype::FdWrite => SubscriptionUnion {
+                    fd_readwrite: unsafe { other.u.fd_readwrite },
+                },
+            },
+        }
+    }
+}
+
+/// The contents of an `event`.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union EventUnion {
+    pub clock: u8,
+    pub fd_readwrite: EventFdReadwrite,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(C)]
+pub struct StackSnapshot {
+    pub user: u64,
+    pub hash: u128,
```

Comment:
FIXME: we are currently using C struct "de/serialization" for a few WASIX syscalls. 

This is mostly fine, but not for `u128`, which has different alignment requirements depending on the platform.
This should be replaced with 32 `u8` , or 2 `u64` or whatever.

Reply:
Now tracked in https://github.com/wasmerio/wasmer/issues/3602